pandas/core/arrays/_mixins.py

```diff
@@ -259,7 +259,15 @@ def shift(self, periods: int = 1, fill_value=None):
     def __setitem__(self, key, value) -> None:
         key = check_array_indexer(self, key)
         value = self._validate_setitem_value(value)
-        self._ndarray[key] = value
+        try:
+            self._ndarray[key] = value
+        except TypeError as exc:
+            # Don't let Python's handling of `complex` make extra complexity for Pandas
+            if self._ndarray.dtype.kind == "c":
+                raise ValueError(
+                    *(x.replace("real", "complex") for x in exc.args)
```

Comment:
why are we tinkering with the exception message?

Reply:
Maybe this is actually a NumPy bug.  When `value` is an object, NumPy appears to try constructing a complex number by passing the object in as the real component, but that, too fails with the message `TypeError: must be a real number, not object`.  But that's a wrong message, because what we really want is to report the failure to construct a complex number.  Here's a test case of sorts:
```
import numpy as np
xx = object()
np.asarray(xx).astype('float64')
# *** TypeError: float() argument must be a string or a real number, not 'object'
np.asarray(xx).astype('complex128')
# *** TypeError: must be real number, not object
```
That's what I'm trying to fix.  So I could file an issue against NumPy, remove the above special case, and Pandas will fix itself when NumPy fixes itself.  Is that the best approach?